// openclaw/dist/plugin-sdk/audio-preflight-BCgj81EE.js
import "./accounts-Dw118nCx.js";
import "./paths-DCNrSyZW.js";
import "./github-copilot-token-Df-R0zCM.js";
import "./config-DonAh6tl.js";
import { $ as logVerbose, nt as shouldLogVerbose } from "./subsystem-D7KkLxSJ.js";
import "./command-format-D4smYdZ1.js";
import "./agent-scope-BY0vP6R-.js";
import "./dock-Dr4CPR_S.js";
import "./message-channel-BhZc-dSE.js";
import "./sessions-BPaGSNZP.js";
import "./plugins-CroA605u.js";
import "./accounts-CWu4D7zt.js";
import "./accounts-DQd1QCaM.js";
import "./bindings-BfMshQTd.js";
import "./paths-BwfwZvKQ.js";
import "./redact-BYwaiynP.js";
import "./errors-BMOVwRE7.js";
import "./path-alias-guards-DF78KPUK.js";
import "./fs-safe-wTNAoqkb.js";
import "./image-ops-BnbZZSRk.js";
import "./ssrf-0Sz9Ifta.js";
import "./fetch-guard-C3cl_qdE.js";
import "./local-roots-CtwY38pl.js";
import "./tool-images-DR4wwgZ6.js";
import { a as resolveMediaAttachmentLocalRoots, n as createMediaAttachmentCache, o as runCapability, r as normalizeMediaAttachments, t as buildProviderRegistry, u as isAudioAttachment } from "./runner-BB45CO83.js";
import "./skills-Bg0y9l5G.js";
import "./chrome-iodTEFVB.js";
import "./store-ByA0o08o.js";
import "./pi-embedded-helpers-W0JmgE4L.js";
import "./thinking-DFrhwk-d.js";
import "./image-DmjyrEGU.js";
import "./pi-model-discovery-DPJRLKS_.js";
import "./api-key-rotation-BZNDbFD8.js";

//#region src/media-understanding/audio-preflight.ts
/**
* Transcribes the first audio attachment BEFORE mention checking.
* This allows voice notes to be processed in group chats with requireMention: true.
* Returns the transcript or undefined if transcription fails or no audio is found.
*/
async function transcribeFirstAudio(params) {
	const { ctx, cfg } = params;
	const audioConfig = cfg.tools?.media?.audio;
	if (!audioConfig || audioConfig.enabled === false) return;
	const attachments = normalizeMediaAttachments(ctx);
	if (!attachments || attachments.length === 0) return;
	const firstAudio = attachments.find((att) => att && isAudioAttachment(att) && !att.alreadyTranscribed);
	if (!firstAudio) return;
	if (shouldLogVerbose()) logVerbose(`audio-preflight: transcribing attachment ${firstAudio.index} for mention check`);
	const providerRegistry = buildProviderRegistry(params.providers);
	const cache = createMediaAttachmentCache(attachments, { localPathRoots: resolveMediaAttachmentLocalRoots({
		cfg,
		ctx
	}) });
	try {
		const result = await runCapability({
			capability: "audio",
			cfg,
			ctx,
			attachments: cache,
			media: attachments,
			agentDir: params.agentDir,
			providerRegistry,
			config: audioConfig,
			activeModel: params.activeModel
		});
		if (!result || result.outputs.length === 0) return;
		const audioOutput = result.outputs.find((output) => output.kind === "audio.transcription");
		if (!audioOutput || !audioOutput.text) return;
		firstAudio.alreadyTranscribed = true;
		if (shouldLogVerbose()) logVerbose(`audio-preflight: transcribed ${audioOutput.text.length} chars from attachment ${firstAudio.index}`);
		return audioOutput.text;
	} catch (err) {
		if (shouldLogVerbose()) logVerbose(`audio-preflight: transcription failed: ${String(err)}`);
		return;
	} finally {
		await cache.cleanup();
	}
}

//#endregion
export { transcribeFirstAudio };
