// openclaw/dist/plugin-sdk/web-Dq3K8te5.js
import { _ as webAuthExists, c as logWebSelfId, d as pickWebChannel, o as WA_WEB_AUTH_DIR } from "./accounts-Dw118nCx.js";
import "./reply-CBCdFV63.js";
import "./paths-DCNrSyZW.js";
import "./github-copilot-token-Df-R0zCM.js";
import "./config-DonAh6tl.js";
import "./subsystem-D7KkLxSJ.js";
import "./command-format-D4smYdZ1.js";
import "./agent-scope-BY0vP6R-.js";
import "./dock-Dr4CPR_S.js";
import "./message-channel-BhZc-dSE.js";
import "./sessions-BPaGSNZP.js";
import "./plugins-CroA605u.js";
import "./accounts-CWu4D7zt.js";
import "./accounts-DQd1QCaM.js";
import "./bindings-BfMshQTd.js";
import "./send-D8rvlR-l.js";
import "./paths-BwfwZvKQ.js";
import "./fetch-DH6qweqj.js";
import "./retry-0h5w0VDJ.js";
import "./redact-BYwaiynP.js";
import "./errors-BMOVwRE7.js";
import "./channel-activity-BMJujSSg.js";
import "./path-alias-guards-DF78KPUK.js";
import "./fs-safe-wTNAoqkb.js";
import "./image-ops-BnbZZSRk.js";
import "./ssrf-0Sz9Ifta.js";
import "./fetch-guard-C3cl_qdE.js";
import "./local-roots-CtwY38pl.js";
import "./ir-CckSHymo.js";
import "./chunk-BwDy78-Y.js";
import "./markdown-tables-B2n9F2UN.js";
import "./render-BkW8s6LR.js";
import "./tables-puYZJwid.js";
import "./send-BYPaz1H6.js";
import "./send-CIjsKh8I.js";
import "./tool-images-DR4wwgZ6.js";
import "./target-errors-BQ4hEy1n.js";
import "./send-Ba1J0e3w.js";
import "./runner-BB45CO83.js";
import "./reply-prefix-xxgZWsi2.js";
import "./tokens-E4MpPrrE.js";
import "./skill-commands-CoCcTOg1.js";
import "./skills-Bg0y9l5G.js";
import "./chrome-iodTEFVB.js";
import "./deliver-CmGxF18F.js";
import "./diagnostic-CBmy3x5W.js";
import "./store-ByA0o08o.js";
import "./pi-embedded-helpers-W0JmgE4L.js";
import { n as monitorWebInbox, t as monitorWebChannel } from "./channel-web-CllPyabo.js";
import "./thinking-DFrhwk-d.js";
import "./image-DmjyrEGU.js";
import "./pi-model-discovery-DPJRLKS_.js";
import "./api-key-rotation-BZNDbFD8.js";
import "./commands-registry-B4lbhSqR.js";
import "./diagnostic-session-state-CT36_PCE.js";
import "./manager-ZZc6U97-.js";
import "./query-expansion-ixGuDYno.js";
import "./send-ZNhdIRbO.js";
import "./outbound-attachment-ht5HtGs3.js";
import "./resolve-route-CbihP70f.js";
import "./proxy-CG7lVglh.js";
import "./replies-BnHUaeWw.js";
import { n as sendMessageWhatsApp } from "./outbound-mEkA6LRD.js";
import { i as waitForWaConnection, t as createWaSocket } from "./session-DKwKBCOV.js";
import { t as loginWeb } from "./login-DFyrFHeL.js";

export { WA_WEB_AUTH_DIR, createWaSocket, logWebSelfId, loginWeb, monitorWebChannel, monitorWebInbox, pickWebChannel, sendMessageWhatsApp, waitForWaConnection, webAuthExists };
