// openclaw/dist/audio-preflight-CxSM9Z3N.js
import "./paths-B4BZAPZh.js";
import { F as shouldLogVerbose, M as logVerbose } from "./utils-BKDT474X.js";
import "./thinking-BB3zi8pq.js";
import "./agent-scope-CTEWT571.js";
import "./subsystem-DypCPrmP.js";
import "./openclaw-root-BjZlrKKM.js";
import "./exec-DNET3cHX.js";
import "./model-selection-DxE_R9Al.js";
import "./github-copilot-token-nncItI8D.js";
import "./boolean-CE7i9tBR.js";
import "./env-CCK0T6mv.js";
import "./host-env-security-CJMD0__Z.js";
import "./env-vars-gp4sxqr7.js";
import "./manifest-registry-CdDckm8W.js";
import "./dock-xjbAbSQ4.js";
import "./message-channel-_kxdCJ2B.js";
import { a as resolveMediaAttachmentLocalRoots, n as createMediaAttachmentCache, o as runCapability, r as normalizeMediaAttachments, s as isAudioAttachment, t as buildProviderRegistry } from "./runner-w1zDalwN.js";
import "./image-B4GTnkwS.js";
import "./models-config-CPRC8TUt.js";
import "./pi-model-discovery-BNjaKkhZ.js";
import "./pi-embedded-helpers-BXFqqNki.js";
import "./sandbox-B1DM6AnX.js";
import "./tool-catalog-BNWFbKiv.js";
import "./chrome-mv_RKsjo.js";
import "./tailscale-H3XFAlm6.js";
import "./tailnet-DhfP1w4L.js";
import "./ws-DWhvxt3w.js";
import "./auth-B01KaQCO.js";
import "./server-context-D0YYteA0.js";
import "./frontmatter-Dn_0bQcK.js";
import "./skills-BLGdpAED.js";
import "./path-alias-guards-DRMUAt9X.js";
import "./paths-CUK9133b.js";
import "./redact-DKstN8W8.js";
import "./errors-CHThjNy0.js";
import "./fs-safe-bED949rs.js";
import "./ssrf-A4NPRuhY.js";
import "./image-ops-BstIV9Yi.js";
import "./store-BTzIaC2w.js";
import "./ports-CqAQp7mK.js";
import "./trash-BgNG7SI6.js";
import "./server-middleware-Gzi4MR3I.js";
import "./sessions-D-LKdTsU.js";
import "./plugins-DpUpOnl4.js";
import "./accounts-BCPiwTlf.js";
import "./accounts-BC4fvRT_.js";
import "./accounts-ClU97yHE.js";
import "./bindings-CIhSyfod.js";
import "./logging-w5jq5901.js";
import "./paths-D6oKH7El.js";
import "./chat-envelope-BZKQmhVe.js";
import "./tool-images-gM5beWVR.js";
import "./tool-display-B99otER_.js";
import "./fetch-guard-BJGpEc3_.js";
import "./api-key-rotation-DMitUtJ-.js";
import "./local-roots-DxHi18nP.js";
import "./model-catalog-BQb1Snsp.js";

//#region src/media-understanding/audio-preflight.ts
/**
* Transcribes the first audio attachment BEFORE mention checking.
* This allows voice notes to be processed in group chats with requireMention: true.
* Returns the transcript or undefined if transcription fails or no audio is found.
*/
async function transcribeFirstAudio(params) {
	const { ctx, cfg } = params;
	const audioConfig = cfg.tools?.media?.audio;
	if (!audioConfig || audioConfig.enabled === false) return;
	const attachments = normalizeMediaAttachments(ctx);
	if (!attachments || attachments.length === 0) return;
	const firstAudio = attachments.find((att) => att && isAudioAttachment(att) && !att.alreadyTranscribed);
	if (!firstAudio) return;
	if (shouldLogVerbose()) logVerbose(`audio-preflight: transcribing attachment ${firstAudio.index} for mention check`);
	const providerRegistry = buildProviderRegistry(params.providers);
	const cache = createMediaAttachmentCache(attachments, { localPathRoots: resolveMediaAttachmentLocalRoots({
		cfg,
		ctx
	}) });
	try {
		const result = await runCapability({
			capability: "audio",
			cfg,
			ctx,
			attachments: cache,
			media: attachments,
			agentDir: params.agentDir,
			providerRegistry,
			config: audioConfig,
			activeModel: params.activeModel
		});
		if (!result || result.outputs.length === 0) return;
		const audioOutput = result.outputs.find((output) => output.kind === "audio.transcription");
		if (!audioOutput || !audioOutput.text) return;
		firstAudio.alreadyTranscribed = true;
		if (shouldLogVerbose()) logVerbose(`audio-preflight: transcribed ${audioOutput.text.length} chars from attachment ${firstAudio.index}`);
		return audioOutput.text;
	} catch (err) {
		if (shouldLogVerbose()) logVerbose(`audio-preflight: transcription failed: ${String(err)}`);
		return;
	} finally {
		await cache.cleanup();
	}
}

//#endregion
export { transcribeFirstAudio };
