// openclaw/dist/audio-preflight-CB9JfXdC.js
import "./agent-scope-CqCg9ftP.js";
import "./paths-B9lBY6-m.js";
import { $ as shouldLogVerbose, X as logVerbose } from "./subsystem-mPRezpll.js";
import "./workspace-uOYF0Vfr.js";
import "./model-selection-DFUO-6cE.js";
import "./github-copilot-token-BgKF-7S1.js";
import "./env-DijpAkAH.js";
import "./boolean-M-esQJt6.js";
import "./dock-Cd2KKvTz.js";
import "./plugins-DZ9-uW3_.js";
import "./accounts-CaRtAz_2.js";
import "./bindings-TdKj4MFr.js";
import "./accounts-8ZWbw0Zq.js";
import "./image-ops-BmyNbawg.js";
import "./pi-model-discovery-hg6Puoxq.js";
import "./message-channel-DKXv9Xa_.js";
import "./pi-embedded-helpers-BKo23cJC.js";
import "./chrome-CXvgb-2a.js";
import "./frontmatter-eHCuq81z.js";
import "./skills-D8DvoIYe.js";
import "./path-alias-guards-4J_nAuNF.js";
import "./redact-B5RjPWCN.js";
import "./errors-BB1m5Yna.js";
import "./fs-safe-BOZhmBSl.js";
import "./ssrf-qU8bat52.js";
import "./store-Rnlfz5g-.js";
import "./sessions-CyA_by-N.js";
import "./accounts-CGJl1Z8Z.js";
import "./paths-BXMwg9Yw.js";
import "./tool-images-Bl0sdAcE.js";
import "./thinking-DjGGQMbY.js";
import "./image-CF2u8kpx.js";
import "./gemini-auth-B6lcpHme.js";
import "./fetch-guard-EUqfaPfm.js";
import "./local-roots-DiuXTFEJ.js";
import { a as resolveMediaAttachmentLocalRoots, n as createMediaAttachmentCache, o as runCapability, r as normalizeMediaAttachments, t as buildProviderRegistry, u as isAudioAttachment } from "./runner-B_Owi3nD.js";

//#region src/media-understanding/audio-preflight.ts
/**
* Transcribes the first audio attachment BEFORE mention checking.
* This allows voice notes to be processed in group chats with requireMention: true.
* Returns the transcript or undefined if transcription fails or no audio is found.
*/
async function transcribeFirstAudio(params) {
	const { ctx, cfg } = params;
	const audioConfig = cfg.tools?.media?.audio;
	if (!audioConfig || audioConfig.enabled === false) return;
	const attachments = normalizeMediaAttachments(ctx);
	if (!attachments || attachments.length === 0) return;
	const firstAudio = attachments.find((att) => att && isAudioAttachment(att) && !att.alreadyTranscribed);
	if (!firstAudio) return;
	if (shouldLogVerbose()) logVerbose(`audio-preflight: transcribing attachment ${firstAudio.index} for mention check`);
	const providerRegistry = buildProviderRegistry(params.providers);
	const cache = createMediaAttachmentCache(attachments, { localPathRoots: resolveMediaAttachmentLocalRoots({
		cfg,
		ctx
	}) });
	try {
		const result = await runCapability({
			capability: "audio",
			cfg,
			ctx,
			attachments: cache,
			media: attachments,
			agentDir: params.agentDir,
			providerRegistry,
			config: audioConfig,
			activeModel: params.activeModel
		});
		if (!result || result.outputs.length === 0) return;
		const audioOutput = result.outputs.find((output) => output.kind === "audio.transcription");
		if (!audioOutput || !audioOutput.text) return;
		firstAudio.alreadyTranscribed = true;
		if (shouldLogVerbose()) logVerbose(`audio-preflight: transcribed ${audioOutput.text.length} chars from attachment ${firstAudio.index}`);
		return audioOutput.text;
	} catch (err) {
		if (shouldLogVerbose()) logVerbose(`audio-preflight: transcription failed: ${String(err)}`);
		return;
	} finally {
		await cache.cleanup();
	}
}

//#endregion
export { transcribeFirstAudio };
