// openclaw/dist/audio-preflight-BQeryxYx.js
import "./paths-B9jPXz5d.js";
import "./subsystem-CDGdCagb.js";
import { P as shouldLogVerbose, j as logVerbose } from "./utils-XeuG5BG2.js";
import "./boolean-DtWR5bt3.js";
import "./auth-profiles-DSWJsDjg.js";
import "./agent-scope-CU2Re8Hu.js";
import "./openclaw-root-BYA-qKQw.js";
import "./exec-QEqvNlTy.js";
import "./github-copilot-token-CHThtPpe.js";
import "./host-env-security-DWcSD4kP.js";
import "./version-DR9Qjj6f.js";
import "./env-vars-tQ4AIdQq.js";
import "./manifest-registry-DdugjusL.js";
import "./dock-CZpRPec8.js";
import "./pi-model-discovery-BqYOh57l.js";
import "./frontmatter-Cqa5i1Xb.js";
import "./skills-DczrXIo7.js";
import "./path-alias-guards-C0SiqlVA.js";
import "./message-channel--ZQegy69.js";
import "./sessions-vBkENF-W.js";
import "./plugins-BqHfIRbv.js";
import "./accounts-x0oZK_GU.js";
import "./accounts-cxEUMdfm.js";
import "./accounts-nu2ApxKu.js";
import "./bindings-ChoDOV8I.js";
import "./logging-BvdokaVt.js";
import "./paths-CFZJ4dSy.js";
import "./chat-envelope-DMwAJAv3.js";
import "./net-DRkWvx5J.js";
import "./tailnet-DXpIDiom.js";
import "./image-ops-D6beOpsg.js";
import "./pi-embedded-helpers-DFCEhnUw.js";
import "./sandbox-BjrYR-KB.js";
import "./tool-catalog-B4gxY3Jd.js";
import "./chrome-DM2YHZd2.js";
import "./tailscale-RsI6sScm.js";
import "./auth-CdRnR6p1.js";
import "./server-context-Cimev08D.js";
import "./paths-DVtIKqPo.js";
import "./redact-Bdn22hWn.js";
import "./errors-DQoYsN9P.js";
import "./fs-safe-HE9JYGe3.js";
import "./ssrf-BOBoLYnV.js";
import "./store-D96YiLuJ.js";
import "./ports-CF8KDGWC.js";
import "./trash-WJ5rbxj0.js";
import "./server-middleware-BkvUFZ1P.js";
import "./tool-images-DsHt1x_L.js";
import "./thinking-lfnd6MLT.js";
import "./models-config-Ce3wYfYf.js";
import "./gemini-auth-CIzdSXC1.js";
import "./fetch-guard-Bho7inTC.js";
import "./local-roots-d9NvRDsE.js";
import "./image-CPvDv8mo.js";
import "./tool-display-B-Kg-QB2.js";
import { a as resolveMediaAttachmentLocalRoots, n as createMediaAttachmentCache, o as runCapability, r as normalizeMediaAttachments, s as isAudioAttachment, t as buildProviderRegistry } from "./runner-B8i872WN.js";
import "./model-catalog-CePudu2Z.js";

//#region src/media-understanding/audio-preflight.ts
/**
* Transcribes the first audio attachment BEFORE mention checking.
* This allows voice notes to be processed in group chats with requireMention: true.
* Returns the transcript or undefined if transcription fails or no audio is found.
*/
async function transcribeFirstAudio(params) {
	const { ctx, cfg } = params;
	const audioConfig = cfg.tools?.media?.audio;
	if (!audioConfig || audioConfig.enabled === false) return;
	const attachments = normalizeMediaAttachments(ctx);
	if (!attachments || attachments.length === 0) return;
	const firstAudio = attachments.find((att) => att && isAudioAttachment(att) && !att.alreadyTranscribed);
	if (!firstAudio) return;
	if (shouldLogVerbose()) logVerbose(`audio-preflight: transcribing attachment ${firstAudio.index} for mention check`);
	const providerRegistry = buildProviderRegistry(params.providers);
	const cache = createMediaAttachmentCache(attachments, { localPathRoots: resolveMediaAttachmentLocalRoots({
		cfg,
		ctx
	}) });
	try {
		const result = await runCapability({
			capability: "audio",
			cfg,
			ctx,
			attachments: cache,
			media: attachments,
			agentDir: params.agentDir,
			providerRegistry,
			config: audioConfig,
			activeModel: params.activeModel
		});
		if (!result || result.outputs.length === 0) return;
		const audioOutput = result.outputs.find((output) => output.kind === "audio.transcription");
		if (!audioOutput || !audioOutput.text) return;
		firstAudio.alreadyTranscribed = true;
		if (shouldLogVerbose()) logVerbose(`audio-preflight: transcribed ${audioOutput.text.length} chars from attachment ${firstAudio.index}`);
		return audioOutput.text;
	} catch (err) {
		if (shouldLogVerbose()) logVerbose(`audio-preflight: transcription failed: ${String(err)}`);
		return;
	} finally {
		await cache.cleanup();
	}
}

//#endregion
export { transcribeFirstAudio };
