// openclaw/dist/tui-cli-DjiT8C20.js
import "./paths-B9jPXz5d.js";
import { f as defaultRuntime } from "./subsystem-CDGdCagb.js";
import { z as theme } from "./utils-XeuG5BG2.js";
import "./boolean-DtWR5bt3.js";
import "./auth-profiles-DSWJsDjg.js";
import "./agent-scope-CU2Re8Hu.js";
import "./openclaw-root-BYA-qKQw.js";
import "./exec-QEqvNlTy.js";
import "./github-copilot-token-CHThtPpe.js";
import "./host-env-security-DWcSD4kP.js";
import "./version-DR9Qjj6f.js";
import "./env-vars-tQ4AIdQq.js";
import "./manifest-registry-DdugjusL.js";
import "./dock-CZpRPec8.js";
import "./frontmatter-Cqa5i1Xb.js";
import "./skills-DczrXIo7.js";
import "./path-alias-guards-C0SiqlVA.js";
import "./message-channel--ZQegy69.js";
import "./sessions-vBkENF-W.js";
import "./plugins-BqHfIRbv.js";
import "./accounts-x0oZK_GU.js";
import "./accounts-cxEUMdfm.js";
import "./accounts-nu2ApxKu.js";
import "./bindings-ChoDOV8I.js";
import "./logging-BvdokaVt.js";
import "./paths-CFZJ4dSy.js";
import "./chat-envelope-DMwAJAv3.js";
import "./client-DZJZEGEG.js";
import "./call-_xu4BCM0.js";
import "./pairing-token-C_35MNZC.js";
import "./net-DRkWvx5J.js";
import "./tailnet-DXpIDiom.js";
import "./image-ops-D6beOpsg.js";
import "./pi-embedded-helpers-DFCEhnUw.js";
import "./sandbox-BjrYR-KB.js";
import "./tool-catalog-B4gxY3Jd.js";
import "./chrome-DM2YHZd2.js";
import "./tailscale-RsI6sScm.js";
import "./auth-CdRnR6p1.js";
import "./server-context-Cimev08D.js";
import "./paths-DVtIKqPo.js";
import "./redact-Bdn22hWn.js";
import "./errors-DQoYsN9P.js";
import "./fs-safe-HE9JYGe3.js";
import "./ssrf-BOBoLYnV.js";
import "./store-D96YiLuJ.js";
import "./ports-CF8KDGWC.js";
import "./trash-WJ5rbxj0.js";
import "./server-middleware-BkvUFZ1P.js";
import "./tool-images-DsHt1x_L.js";
import "./thinking-lfnd6MLT.js";
import "./commands-DjmLO4aZ.js";
import "./commands-registry-Ce_dY54o.js";
import "./tool-display-B-Kg-QB2.js";
import { t as parseTimeoutMs } from "./parse-timeout-DmEcHIsV.js";
import { t as formatDocsLink } from "./links-DMmQqqmC.js";
import { t as runTui } from "./tui-DFTtVWkH.js";

//#region src/cli/tui-cli.ts
function registerTuiCli(program) {
	program.command("tui").description("Open a terminal UI connected to the Gateway").option("--url <url>", "Gateway WebSocket URL (defaults to gateway.remote.url when configured)").option("--token <token>", "Gateway token (if required)").option("--password <password>", "Gateway password (if required)").option("--session <key>", "Session key (default: \"main\", or \"global\" when scope is global)").option("--deliver", "Deliver assistant replies", false).option("--thinking <level>", "Thinking level override").option("--message <text>", "Send an initial message after connecting").option("--timeout-ms <ms>", "Agent timeout in ms (defaults to agents.defaults.timeoutSeconds)").option("--history-limit <n>", "History entries to load", "200").addHelpText("after", () => `\n${theme.muted("Docs:")} ${formatDocsLink("/cli/tui", "docs.openclaw.ai/cli/tui")}\n`).action(async (opts) => {
		try {
			const timeoutMs = parseTimeoutMs(opts.timeoutMs);
			if (opts.timeoutMs !== void 0 && timeoutMs === void 0) defaultRuntime.error(`warning: invalid --timeout-ms "${String(opts.timeoutMs)}"; ignoring`);
			const historyLimit = Number.parseInt(String(opts.historyLimit ?? "200"), 10);
			await runTui({
				url: opts.url,
				token: opts.token,
				password: opts.password,
				session: opts.session,
				deliver: Boolean(opts.deliver),
				thinking: opts.thinking,
				message: opts.message,
				timeoutMs,
				historyLimit: Number.isNaN(historyLimit) ? void 0 : historyLimit
			});
		} catch (err) {
			defaultRuntime.error(String(err));
			defaultRuntime.exit(1);
		}
	});
}

//#endregion
export { registerTuiCli };
